refactor: correct `callbacks` type

yyueniao opened the pull request:
> _NOTE_:
>
> - It's a good idea to open an issue first to discuss potential changes.
> - Please make sure that you are _NOT_ opening a PR to fix a potential security vulnerability. Instead, please follow the [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md) to disclose the issue to us confidentially.

## ☕️ Reasoning

Some of the methods in `callbacks` have a wrong params type. There are some of them should be optional.

## 🧢 Checklist

- [x] Documentation
- [x] Tests
- [x] Ready to be merged

## 🎫 Affected issues

Fixes: https://github.com/nextauthjs/next-auth/discussions/3951 (previous https://github.com/nextauthjs/next-auth/issues/3577)

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

yyueniao committed:
refactor: correct `callbacks` type

Some of the methods in `callbacks` have a wrong params type. There are
some of them should be optional.

Juneezee commented on packages/core/src/types.ts:
```diff
@@ -178,6 +178,101 @@ export interface Profile {
+   * Otherwise, it will be the full {@link JWT} for subsequent calls.
+   */
+  token: JWT
+  /**
+   * Either the result of the {@link OAuthConfig.profile} or the {@link CredentialsConfig.authorize} callback.
+   * @note available when `trigger` is `"signIn"` or `"signUp"`.
+   *
+   * Resources:
+   * - [Credentials Provider](https://authjs.dev/reference/core/providers_credentials)
+   * - [User database model](https://authjs.dev/reference/adapters#user)
+   */
+  user?: User | AdapterUser
+  /**
+   * Contains information about the provider that was used to sign in.
+   * Also includes {@link TokenSet}
+   * @note available when `trigger` is `"signIn"` or `"signUp"`
+   */
+  account?: A | null
```
One drawback of making `user` and `account` optional is that we will have to check them for `undefined` even when `trigger` is `signIn` or `signUp`. This is a breaking change for existing TypeScript users.

Before this PR:

```typescript
callbacks: {
  async jwt({ token, user, account, trigger }) {
    if (trigger === "signIn") {
      console.log(user); // user is not User | AdapterUser
      console.log(account); // account is A | null
    }
  }
}
```

After this PR:

```typescript
callbacks: {
  async jwt({ token, user, account, trigger }) {
    if (trigger === "signIn") {
      console.log(user); // user is User | AdapterUser | undefined
      console.log(account); // account is A | null | undefined
    }
  }
}
```

yyueniao replied:
@Juneezee I use type union to help typescript developers [narrowing down](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions) the types. Unfortunately, this makes jsdocs not DRY anymore. Do you have any advice?

yyueniao commented on packages/core/src/types.ts:
```diff
@@ -379,7 +385,7 @@ export interface EventCallbacks {
    * - `session`: The session object from your adapter.
    */
-  session: (message: { session: Session; token: JWT }) => Awaitable<void>
+  session: (message: { session: Session; token?: JWT }) => Awaitable<void>
```
When I search through this project, there are only two references in this project, and both of them provide `session`. So at least at this point `session` is not optional.

yyueniao committed:
refactor: improve type hint

Use type union to improve developer experience when using this
library with typescript.

Ref: https://github.com/nextauthjs/next-auth/pull/8404#discussion_r1306824424
